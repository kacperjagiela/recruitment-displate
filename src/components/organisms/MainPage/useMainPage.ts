import { useEffect, useState } from 'react';
import { debounce } from 'throttle-debounce';

import Dog from '~/types/Dog';
import { filterDogsByBread } from '~/utils/filterDogsByBread';

interface UseMainPage {
    currentDog: Dog;
    onDogButtonClick: (dog: Dog) => void;
    isImageLoading: boolean;
    onImageLoad: () => void;
    searchValue: string;
    onSearch: (searchValue: string) => void;
    filteredDogs: Dog[];
}

const useMainPage = (dogs: Dog[]): UseMainPage => {
    const [currentDog, setCurrentDog] = useState({ name: '', url: '' });
    const [searchValue, setSearchValue] = useState('');
    const [isImageLoading, setIsImageLoading] = useState(true);

    const [filteredDogs, setFilteredDogs] = useState(dogs);

    useEffect(() => {
        if (searchValue.length > 0) {
            filterDogsBySearchValue(searchValue);
        } else {
            setFilteredDogs(dogs);
        }
    }, [searchValue]);

    const onDogButtonClick = (dog: Dog) => {
        setIsImageLoading(true);
        setCurrentDog(dog);
    };

    const filterDogsBySearchValue = debounce(350, (searchValue) => {
        const matchedDogs = dogs.filter((dog) => filterDogsByBread(searchValue, dog.name));

        setFilteredDogs(matchedDogs);
    });

    const onImageLoad = () => setIsImageLoading(false);

    const onSearch = (searchValue: string) => setSearchValue(searchValue);

    return {
        currentDog,
        onDogButtonClick,
        isImageLoading,
        onImageLoad,
        searchValue,
        onSearch,
        filteredDogs,
    };
};

export default useMainPage;
