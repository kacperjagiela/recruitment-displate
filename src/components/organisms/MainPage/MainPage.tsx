import { useDisclosure } from '@chakra-ui/hooks';
import { Box } from '@chakra-ui/layout';
import {
    Modal,
    ModalBody,
    ModalCloseButton,
    ModalContent,
    ModalFooter,
    ModalHeader,
    ModalOverlay,
} from '@chakra-ui/modal';
import { Stack } from '@chakra-ui/react';

import { Button } from '~/components/atoms';
import { DogsList } from '~/components/molecules';
import Dog from '~/types/Dog';

import useMainPage from './useMainPage';

interface Props {
    dogs: Dog[];
}

const MainPage: React.FC<Props> = ({ dogs }: Props) => {
    const { isOpen, onOpen, onClose } = useDisclosure();
    const { currentDog, onDogButtonClick } = useMainPage();

    return (
        <Box bg="blue.900" px={['10%', '20%']}>
            <DogsList dogs={dogs} onModalOpen={onOpen} onDogButtonClick={onDogButtonClick} />

            <Modal isOpen={isOpen} onClose={onClose}>
                <ModalOverlay />
                <ModalContent>
                    <ModalHeader>{currentDog.name}</ModalHeader>
                    <ModalCloseButton />
                    <ModalBody>{currentDog.url}</ModalBody>

                    <ModalFooter>
                        <Stack spacing={3} direction="row">
                            <Button buttonText="Another pupper" onClick={onClose} />
                            <Button buttonText="Close" variant="outline" onClick={onClose} />
                        </Stack>
                    </ModalFooter>
                </ModalContent>
            </Modal>
        </Box>
    );
};

export default MainPage;
