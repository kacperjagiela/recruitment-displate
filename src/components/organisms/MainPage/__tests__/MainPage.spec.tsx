import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';

import MainPage from '../MainPage';

describe('<MainPage />', () => {
    it('renders correctly', () => {
        const { container } = render(<MainPage dogs={[{ name: 'testdoge', url: 'https://displate.com' }]} />);

        expect(container).toMatchSnapshot();
    });

    it('opens modal on dog button click', () => {
        const { getByText } = render(<MainPage dogs={[{ name: 'testdoge', url: 'https://displate.com' }]} />);

        waitFor(() => {
            fireEvent.click(getByText('testdoge'));

            expect(getByText('Another pupper')).toBeInTheDocument();
        });
    });

    it('closes modal on button click', () => {
        const { getByText } = render(<MainPage dogs={[{ name: 'testdoge', url: 'https://displate.com' }]} />);

        waitFor(() => {
            fireEvent.click(getByText('testdoge'));

            const closeButton = getByText('Close');

            expect(closeButton).toBeInTheDocument();

            fireEvent.click(closeButton);

            expect(closeButton).not.toBeInTheDocument();
        });
    });
});
