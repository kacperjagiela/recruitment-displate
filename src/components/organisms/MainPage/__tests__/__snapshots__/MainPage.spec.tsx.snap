// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainPage /> renders correctly 1`] = `
<div>
  <div
    class="css-p8m5bp"
  >
    <div
      class="css-1k5s6ga"
    >
      <div
        class="css-18ws1q3"
      >
        <h2
          class="chakra-text css-0"
        >
          Woof Woof Gallery
        </h2>
        <h3
          class="chakra-text css-o5g2qp"
        >
          Already know the breed?
        </h3>
        <h4
          class="chakra-text css-0"
        >
          Search for it
        </h4>
        <div
          class="chakra-input__group css-4302v8"
        >
          <input
            class="chakra-input css-0"
            placeholder="Germanshepherd.."
            value=""
          />
        </div>
      </div>
      <div
        class="chakra-stack css-1tobaed"
      >
        <button
          class="chakra-button css-omiz5g"
          type="button"
        >
          testdoge
        </button>
      </div>
    </div>
    <h5
      class="chakra-text css-6uhn4g"
    >
      Developed by
       
      <a
        class="chakra-link css-1jk86wo"
        href="https://github.com/kacperjagiela"
      >
        Kacper Jagie≈Ça
      </a>
       
      for
       
      <a
        class="chakra-link css-1jk86wo"
        href="https://displate.com"
      >
        Displate
      </a>
    </h5>
  </div>
</div>
`;
